src/contracting/storage/orm.py

```diff
@@ -138,4 +148,127 @@ def __getitem__(self, item):
+        self._params = params
+        self._contract = contract
+        self._event = event
+
+        assert isinstance(params, dict), "Args must be a dictionary."
+        assert len(params) > 0, "Args must have at least one argument."
+        # Check for indexed arguments with a maximum of three
+        indexed_args_count = sum(1 for arg in params.values() if arg.get("idx", False))
+        assert (
+            indexed_args_count <= 3
+        ), "Args must have at most three indexed arguments."
+        for param in params.values():
+            if not isinstance(param["type"], tuple):
+                param["type"] = (param["type"],)
+
+            assert all(
+                issubclass(t, (str, int, float, bool))
+                for t in param["type"]
+            ), "Each type in args must be str, int, float, or bool."
+
+
+    def write_event(self, event_data):
+        print(f"event_data: {event_data}")
+        assert len(event_data) == len(
+            self._params
+        ), "Event Data must have the same number of arguments as specified in the event."
+
+        # Check for unexpected arguments
+        for arg in event_data:
+            assert (
+                arg in self._params
+            ), f"Unexpected argument {arg} in the data dictionary."
+
+        # Check for missing and type-mismatched arguments
+        for arg in self._params:
+            assert (
+                arg in event_data
+            ), f"Argument {arg} is missing from the data dictionary."
+
+            # Check the type of the argument
+            assert isinstance(event_data[arg], self._params[arg]["type"]), (
+                f"Argument {arg} is the wrong type! "
+                f"Expected {self._params[arg]['type']}, got {type(event_data[arg])}."
+            )
+
+            # Check the size of the argument
+            value_size = len(str(event_data[arg]).encode("utf-8"))
+            assert (
+                value_size <= 1024
+            ), f"Argument {arg} is too large ({value_size} bytes). Max is 1024 bytes."
+        
+        event = {
+            "contract": self._contract,
+            "event": self._event,
+            "signer": rt.context.signer,
+            "caller": rt.context.caller,
+            "data_indexed": {
+                arg: event_data[arg]
+                for arg in self._params
+                if self._params[arg].get("idx", False)
+            },
+            "data": {
+                arg: event_data[arg]
+                for arg in self._params
+                if not self._params[arg].get("idx", False)
+            },
+        }
+
+
+        for arg, value in event["data_indexed"].items():
+            assert isinstance(
+                value, self._params[arg]["type"]
+            ), f"Indexed argument {arg} is the wrong type! Expected {self._params[arg]['type']}, got {type(value)}."
+            encoded = encode_kv(arg, value)
+            rt.deduct_write(*encoded, multiplier=0.5)
```

Comment:
writes here are 1/2 the price of writes to abci_state. reasoning here is it will take up less space as it's being indexed by cometbft / bds. 0.6 or so is probably more appropriate.

Reply:
Deducting reads is not working on testnet for this feature right now. unclear why.